fat/fatfs.8
.Dd April 26, 2025
.Dt FATFS 8
.Os
.Sh NAME
.Nm fatfs
.Nd fat filesystem driver
.Sh SYNOPSIS
.Nm
.Op Fl bf
.Op Fl o Ar mount-options
.Op Fl p Ar path
.Ar device
.Ar mountpoint
.Sh DESCRIPTION
.Nm
mounts the FAT filesystem on the
.Ar device
for reading and writing at the
.Ar mountpoint .
.Nm
forks and runs in the background once the filesystem has been mounted.
.Pp
A compatible FAT filesystem can be created using
.Xr mkfs.fat 8 ,
and such filesystems can be checked and repaired using
.Xr fsck.fat 8 .
The filesystem label can be set with
.Xr fatlabel 8 .
Filesystems can be automatically mounted by
.Xr init 8
on boot by creating an entry in
.Xr fstab 5
with the
.Sy fat
or
.Sy efi
type.
.Xr disked 8
is typically used for partition and filesystem administration.
The
.Sy efi
filesystem type is the same as the
.Sy fat
filesystem, except it uses a different
.Xr mbr 7
id and a different
.Xr guid 7
guid to identify the partition as the EFI System Partition.
.Pp
The FAT filesystem uses a File Allocation Table to allocate clusters.
The block device is split into fixed sized clusters.
The FAT entries are either 12, 16, or 32 bit wide, solely depending on the
number of clusters and what width is needed.
Files and directories are stored as a pointer to the first cluster, and the FAT
is used as a singly linked list to next cluster.
A backup FAT is commonly formatted in the filesystem for data recovery purposes.
The FAT filesystem has exactly the same features regardless of the FAT entry
width.
.Pp
The FAT filesystem supports only files and directories.
Hard links, symbolic links, block devices, sockets, fifos, block devices,
character devices, and any other special files are not possible.
Owner and group metadata does not exist.
The only permission metadata is an owner write bit.
There are no read and execute permission bits.
The access timestamp has day precision.
The modified timestamp has 2 second precision.
The creation timestamp has 10 ms precision, but is not accessible through
.Xr stat 2
and
.Xr utimens 2 .
There is no status change timestamp and the modified time is used instead.
All timestamps store the year as an unsigned 7-bit value since 1980 and will
overflow in the year 2108.
Directory entries are assigned a 8.3 short name, with up to 8 uppercase
characters for the file name and up to 3 uppercase characters for the file
extension.
The original name is used as a short name if it already is in the 8.3 format.
Otherwise a similar short will be used (with a prefix of file name, a
.Sq ~
character, numeric digits to disambiguate similar files, and three characters
from the file extensions), and a long file name entry will be
added with the original file name.
Long file names can store up to 255 UTF-16 code units.
.Pp
The options are as follows:
.Bl -tag -width "12345678"
.It Fl b , \-background
Run in the background in a subprocess after successfully mounting.
This is the default behavior.
.It Fl f , \-foreground
Run in the foreground and do not detach in a subprocess.
.It Fl o Ar mount-options
Mount the filesystem with the following comma-separated
.Ar mount-options :
.Bl -tag -width "1234567890"
.It Sy cache Ns = Ns Ar size
Sets the filesystem cache to the specified
.Ar size ,
which is measured in an optional suffix:
.Sy %
for percent of system memory,
.Sy K
for KiB,
.Sy M
for MiB,
or
.Sy G
for GiB.
The default is
.Sy 10%
of the system memory.
.It Sy ro
Mount the filesystem as read-only.
.It Sy rw
Mount the filesystem for read and write.
This option is the default behavior.
.El
.It Fl p , \-pretend-mount-path Ns = Ns Ar path
When answering requests about where the filesystem is mounted, reply with this
.Pa path
instead of the actual mountpoint in
.Xr tcgetblob 2 .
This behavior is useful for
.Xr chroot 2
environments.
.El
.Pp
The following identifiers are available to recognize a filesystem in
.Xr fstab 5
in the
.Ft fs_spec
field and can be queried with the
.Xr disked 8
.Sy id
command:
.Bl -tag -width "12345678"
.It Sy LABEL Ns = Ns Ar label
Search for a filesystem with this
.Ar label
as the volume label.
The volume label can be changed with
.Xr fatlabel 8 .
.It Sy SERIAL Ns = Ns Ar serial
Search for a filesystem with this
.Ar serial
number as the volume id in the
.Sy XXXX-XXXX
upper-case hexadecimal format.
.It Sy UUID Ns = Ns Ar uuid
Search for a filesystem with this
.Ar uuid ,
containing the 4 byte file serial number and 8 byte volume label and a trailing
nul byte.
This identifier is recommended for filesystem searches to avoid mounting the
wrong filesystem, although it might not be globally unique.
Note that if the volume label is changed with
.Xr fatlabel 8 ,
then
.Xr fstab 5
must be updated with the new
.Sy UUID .
.El
.Sh IMPLEMENTATION NOTES
If
.Fl f ,
then
.Nm
signals readiness when it has successfully mounted the filesystem.
.Pp
FAT is a case insensitive filesystem, but
.Nm
implements case sensitivity.
Short names will always be upper case, but the original case will be preserved
in the long name.
.Nm
allows creating a file as long as its long name is different bytewise from any
other long name in the directory.
.Nm
does not allow opening a file by its short name, unless it has no long name, in
which case the short name is used as its real name.
If the requested name is in the 8.3 format, but a file already uses that short
name, but the long name is not used, then the requested name is used as the long
name and an unused short name is allocated.
Other operating systems with case insensitive FAT filesystems will not
interoperate if such files are created and such files may not be accessible
except through their assigned short names.
.Pp
FAT officially stores timestamps in the local timezone, without storing what
timezone that is, but
.Nm
uses UTC instead to store time losslessly.
This behavior may cause time differences when interoperating with other
implementations.
.Pp
The Windows FAT filesystem driver limits directories to 65536 entries.
.Nm
has no such limitation, except it will only search for up to the ~65536 suffix
when allocating an unused short name.
.Pp
The first cluster of a file is used as its inode number.
If a zero size file without a first cluster is encountered, it will be allocated
a first cluster, so it has an unique and permanent inode number.
.Xr fsck.fat 8
does not like this behavior and will undo it.
.Pp
The filesystem is marked as dirty in FAT entry 1 when mounted.
This behavior happens even on FAT12, even though it is not an official feature,
but is done since running
.Xr fsck.fat 8
is essential whenever the filesystem is damaged or not cleanly unmounted.
.Sh ASYNCHRONOUS EVENTS
.Bl -tag -width "SIGTERM"
.It Dv SIGTERM
Request daemon termination.
.Nm
will exit after gracefully unmounting the filesystem.
.El
.Sh EXIT STATUS
If
.Fl b ,
.Nm
exits 0 after successfully mounting the filesystem and serving requests in a
background process.
.Pp
If
.Fl f ,
.Nm
signals readiness on the
.Ev READYFD
file descriptor when filesystem has been mounted, and continues running in the
same process.
.Pp
.Nm
runs as a
.Xr daemon 7
until stopped by
.Dv SIGTERM
or until the filesystem is unmounted with
.Xr unmount 8
or
.Xr unmount 2 ,
after which
.Nm
exits 0 if the filesystem was cleanly unmounted.
.Pp
.Nm
exits non-zero on fatal errors.
.Sh EXAMPLES
.Bd -literal
$ mkfs.fat /dev/foo0
$ fatfs /dev/foo0 /mnt
$ echo bar > /mnt/bar
$ unmount /mnt
$ fsck.fat /dev/foo0
.Ed
.Sh SEE ALSO
.Xr fstab 5 ,
.Xr disked 8 ,
.Xr fatlabel 8 ,
.Xr fsck.fat 8 ,
.Xr init 8 ,
.Xr lsblk 8 ,
.Xr mkfs.fat 8 ,
.Xr unmount 8
.Sh BUGS
.Xr fsync 2
does not sync the inode data contents, but only its metadata.
.Nm
will sync in the background, but this fact cannot be observed yet.
To reliably sync, unmount the filesystem.
.Pp
libmount uses the volume label in the first sector rather than the volume label
in the root directory.
.Pp
Timestamps store years as an unsigned 7-bit value since 1980 and will overflow
in the year 2108.
.Pp
Reading a directory takes O(n^2) quadratic time due to limitations in the
user-space filesystem protocol.


fat/fatfs.cpp
/*
 * Copyright (c) 2013, 2014, 2015, 2016, 2023, 2025 Jonas 'Sortie' Termansen.
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 *
 * fatfs.cpp
 * The File Allocation Table (FAT) filesystem.
 */

#include <sys/stat.h>
#include <sys/types.h>

#include <assert.h>
#include <dirent.h>
#include <endian.h>
#include <err.h>
#include <errno.h>
#include <fcntl.h>
#include <getopt.h>
#include <inttypes.h>
#include <locale.h>
#include <stddef.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>

#if defined(__sortix__)
#include "fsmarshall.h"
#else
#include "fuse.h"
#endif

#include "block.h"
#include "device.h"
#include "fat.h"
#include "fatfs.h"
#include "filesystem.h"
#include "inode.h"
#include "ioleast.h"
#include "util.h"

uid_t request_uid;
uid_t request_gid;

int main(int argc, char* argv[])
{
	setlocale(LC_ALL, "");

	size_t memory;
#ifdef __sortix__
	memstat(NULL, &memory);
#else
	memory = (uintmax_t) sysconf(_SC_PAGE_SIZE) *
	         (uintmax_t) sysconf(_SC_PHYS_PAGES);
#endif
	size_t cache_size = (memory / 10);

	const char* fuse_options = NULL;
	const char* pretend_mount_path = NULL;
	bool foreground = false;
	bool write = true;
	enum
	{
		OPT_FUSE_OPTIONS = 257,
	};
	const struct option longopts[] =
	{
		{"fuse-options", required_argument, NULL, OPT_FUSE_OPTIONS},
		{"background", no_argument, NULL, 'b'},
		{"foreground", no_argument, NULL, 'f'},
		{"pretend-mount-path", required_argument, NULL, 'p'},
		{0, 0, 0, 0}
	};
	const char* opts = "bfo:p:";
	int opt;
	while ( (opt = getopt_long(argc, argv, opts, longopts, NULL)) != -1 )
	{
		switch ( opt )
		{
		case OPT_FUSE_OPTIONS: fuse_options = optarg; break;
		case 'b': foreground = false; break;
		case 'f': foreground = true; break;
		case 'o':
		{
			char* arg = optarg;
			char* save;
			char* tok;
			while ( (tok = strtok_r(arg, ",", &save)) )
			{
				if ( !strcmp(tok, "ro") )
					write = false;
				else if ( !strcmp(tok, "rw") )
					write = true;
				else if ( !strncmp(tok, "cache=", strlen("cache=")) )
				{
					char* end;
					uintmax_t val = strtoumax(tok + strlen("cache="), &end, 10);
					uintmax_t mult;
					if ( !strcmp(end, "%" ) )
						mult = (memory / 100);
					else if ( !strcmp(end, "K" ) )
						mult = 1024ULL << 0;
					else if ( !strcmp(end, "M" ) )
						mult = 1024ULL << 10;
					else if ( !strcmp(end, "G" ) )
						mult = 1024ULL << 20;
					else if ( !strcmp(end, "" ) )
						mult = 1;
					else
						errx(1, "warning: invalid cache size: %s", tok);
					if ( __builtin_mul_overflow(val, mult, &cache_size) )
						errx(1, "warning: invalid cache size: %s", tok);
				}
				else
					warnx("warning: unknown mount option: %s", tok);
				arg = NULL;
			}
			break;
		}
		case 'p': pretend_mount_path = optarg; break;
		default: return 1;
		}
	}

	if ( argc - optind < 1 )
		errx(1, "expected device");
	if ( argc - optind < 2 )
		errx(1, "expected mountpoint");

	const char* device_path = argv[optind + 0];
	const char* mount_path = argv[optind + 1];

	if ( !pretend_mount_path )
		pretend_mount_path = mount_path;

	int fd = open(device_path, write ? O_RDWR : O_RDONLY);
	if ( fd < 0 )
		err(1, "%s", device_path);

	struct stat st;
	fstat(fd, &st);

	// Read the bios parameter block from the filesystem so we can verify it.
	struct fat_bpb bpb;
	if ( preadall(fd, &bpb, sizeof(bpb), 0) != sizeof(bpb) )
	{
		if ( errno == EEOF )
			errx(1, "%s: Isn't a FAT filesystem (too short)", device_path);
		else
			err(1, "read: %s", device_path);
	}

	// Verify the boot signature.
	if ( !(bpb.boot_signature[0] == 0x55 && bpb.boot_signature[1] == 0xAA) )
		errx(1, "%s: Isn't a FAT filesystem (no boot signature)", device_path);
	// Verify the jump instruction at the start of the boot sector.
	if ( !(bpb.jump[0] == 0xEB && bpb.jump[2] == 0x90) &&
	     !(bpb.jump[0] == 0xE9) )
		errx(1, "%s: Isn't a FAT filesystem (bad jump)", device_path);
	// Verify the sector size.
	uint16_t bytes_per_sector =
		bpb.bytes_per_sector_low | bpb.bytes_per_sector_high << 8;
	if ( bytes_per_sector < 512 ||
	     (bytes_per_sector & (bytes_per_sector - 1)) ||
	     4096 < bytes_per_sector )
		errx(1, "%s: Bad number of bytes per sector: %u",
		     device_path, bytes_per_sector);
	// Verify the size of the root directory.
	uint16_t root_dirent_count =
		bpb.root_dirent_count_low | bpb.root_dirent_count_high << 8;
	uint32_t root_dir_sectors =
		divup<uint32_t>(root_dirent_count * sizeof(struct fat_dirent),
		                bytes_per_sector);
	if ( (root_dirent_count * 32) % bytes_per_sector )
		errx(1, "%s: Root directory entry size (%u) is not sector aligned (%u)",
		     device_path, root_dirent_count * 32, bytes_per_sector);
	// Verify the number of sectors per FAT.
	uint32_t sectors_per_fat =
		le16toh(bpb.sectors_per_fat) ? le16toh(bpb.sectors_per_fat) :
	    le32toh(bpb.fat32_sectors_per_fat);
	if ( !sectors_per_fat )
		errx(1, "%s: Bad number of sectors per fat: %u",
		     device_path, sectors_per_fat);
	// Verify the total number of sectors.
	uint32_t total_sectors =
		bpb.total_sectors_low | bpb.total_sectors_high << 8;
	if ( !total_sectors )
		total_sectors = le32toh(bpb.total_sectors_large);
	if ( !total_sectors )
		errx(1, "%s: Bad total number of sectors: %u",
		     device_path, total_sectors);
	if ( st.st_size / bytes_per_sector < total_sectors )
		errx(1, "%s: Device has fewer sectors (%u) than filesystem (%u)",
		     device_path, (uint32_t) (st.st_size / bytes_per_sector),
		     total_sectors);
	// Verify the FAT count is non-zero.
	if ( bpb.fat_count < 1 )
		errx(1, "%s: Bad fat count: %u", device_path, bpb.fat_count);
	// Verify the FAT offset is beyond the BPB.
	if ( !le16toh(bpb.reserved_sectors) )
		errx(1, "%s: Bad reserved sector count: %u",
		     device_path, le16toh(bpb.reserved_sectors));
	// Verify the number of sectors for the FAT.
	uint32_t fat_sectors;
	if ( __builtin_mul_overflow(bpb.fat_count, sectors_per_fat, &fat_sectors) )
		errx(1, "%s: Bad fat size: %u * %u",
		     device_path, bpb.fat_count, sectors_per_fat);
	// Verify the sector where the data begins.
	uint32_t data_offset =
		le16toh(bpb.reserved_sectors) + fat_sectors + root_dir_sectors;
	if ( data_offset > total_sectors )
		errx(1, "%s: Bad data offset (%u) is greater than sector total (%u)",
		     device_path, data_offset, total_sectors);
	uint32_t data_sectors = total_sectors - data_offset;
	// Verify the sectors per cluster is a non-zero power of two.
	if ( bpb.sectors_per_cluster == 0 ||
	     (bpb.sectors_per_cluster & (bpb.sectors_per_cluster - 1)) )
		errx(1, "%s: Bad number of sectors per cluster: %u",
		     device_path, bpb.sectors_per_cluster);
	// Verify the number ofc clusters.
	uint32_t cluster_count = data_sectors / bpb.sectors_per_cluster;
	if ( cluster_count < 1 || 0xFFFFFF7 - 2 <= cluster_count )
		errx(1, "%s: Bad number of clusters: %u", device_path, cluster_count);
	// Determine the FAT type.
	uint8_t fat_type =
		cluster_count < 4085 ? 12 : cluster_count < 65525 ? 16 : 32;
	// Verify the FAT has enough clusters.
	uint64_t fat_cluster_count =
		((uint64_t) sectors_per_fat * bytes_per_sector * 8) / fat_type;
	if ( fat_cluster_count < 2 + cluster_count )
		errx(1, "%s: FAT doesn't have enough clusters: %u < %u",
		     device_path, (uint32_t) fat_cluster_count, 2 + cluster_count);
	// Verify the root directory isn't empty on FAT12/16.
	if ( root_dirent_count < 1 && fat_type < 32 )
		errx(1, "%s: Bad root directory entries: %u",
		     device_path, bytes_per_sector);
	// Verify the filesystem version.
	if ( fat_type == 32 && le16toh(bpb.fat32_version) != 0x0000 )
		errx(1, "%s: Unsupported filesystem version 0x%04x", device_path,
		     le16toh(bpb.fat32_version));
	// Verify the root directory cluster is within bounds.
	if ( fat_type == 32 &&
	     (le32toh(bpb.fat32_root_cluster) < 2 ||
	      2 + cluster_count < le32toh(bpb.fat32_root_cluster)) )
		errx(1, "%s: Bad root directory cluister: %u",
		     device_path, le32toh(bpb.fat32_root_cluster));
	// Verify fsinfo is within bounds.
	if ( fat_type == 32 &&
	     (le16toh(bpb.fat32_fsinfo) < 1 ||
	      le16toh(bpb.reserved_sectors) <= le16toh(bpb.fat32_fsinfo)) )
		errx(1, "%s: Bad fsinfo sector: %u",
		     device_path, le16toh(bpb.fat32_fsinfo));

	size_t block_limit = cache_size / bytes_per_sector;

	// TODO: The FAT and clusters are not aligned to cluster size so
	//       we can't use the cluster size here. Perhaps refactor the
	//       device so we can deal with whole clusters.
	Device* dev = new Device(fd, device_path, bytes_per_sector, block_limit,
	                         write);
	if ( !dev ) // TODO: Use operator new nothrow!
		err(1, "malloc");
	Block* bpb_block = dev->GetBlock(0);
	if ( !bpb_block )
		err(1, "Reading bpb block");
	Filesystem* fs = new Filesystem(dev, pretend_mount_path, bpb_block);
	if ( !fs ) // TODO: Use operator new nothrow!
		err(1, "malloc");
	if ( !fs->WasUnmountedCleanly() )
		warnx("warning: %s: Filesystem wasn't unmounted cleanly", device_path);
	if ( write && !fs->MarkMounted() )
		err(1, "failed to mark filesystem as mounted");
	if ( !(fs->root = fs->CreateInode(fs->root_inode_id, NULL, NULL, NULL)) )
		err(1, "opening /");

	if ( !mount_path )
		return 0;

#if defined(__sortix__)
	(void) fuse_options;
	return fsmarshall_main(argv[0], mount_path, foreground, fs, dev);
#else
	return fat_fuse_main(argv[0], mount_path, fuse_options, foreground, fs,
	                     dev);
#endif
}
